src/ddmd/dcache.d

```diff
@@ -0,0 +1,278 @@
+module ddmd.dcache;
+
+import core.sys.posix.sys.mman;
+import core.sys.posix.fcntl;
+import core.sys.posix.semaphore;
+import core.sys.posix.time;
+import core.sys.posix.unistd;
+import core.stdc.errno;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+version(Posix)
+{
+        
+    struct DCache
+    {
+        /// Load the cache identified by 'keyspace'
+        void initialize(const(char)[] keyspace)
+        {
+            char[512] buf;
+            const(char)* home = getenv("HOME");
+            if (home == null) home = "/tmp";
```

Comment:
This should not be dependent on environmental variables, and you should not hardcode `/tmp`.

Adding a command line switch to control the cache directory would be better.  Or fallback to the  build directory.

Reply:
I thought using .something in a HOME directory was super common for build tools of all sorts.

Regardless "build" directory is fairly bad choice - in this case you cannot reuse cache across different builds, the point is you can and likely should.

Reply:
If you have persistent storage across builds then you can just pass `-flag=/dir` in compilation.

Dub might be able to leverage this too, with a per project cache.

Reply:
> If you have persistent storage across builds then you can just pass -flag=/dir in compilation.

Or you could have a sensible default - use global cache (per compiler version string). There is no harm done.

Why the most useful case must be hidden behind a flag (that you can easily get wrong)?

Reply:
The compiler shouldn't depend on environmental state.  Control should be given to the build tool over where cached modules are written.

Reply:
And how defaulting to build directory accomplishes that?

I’ll redo -cache switch to allow optional path to cache file. We can then discuss where the default should go.

-cache=mmap:some/path/here

Want control - pass an option, don’t want control - don’t pass the path.

Regardless - shared cache is an environmental state, in fact, it can live on a separate host. If you just wanted to have an option to pass it explicitly, I guess I’m fine with that.

Reply:
> And how defaulting to build directory accomplishes that?

`dmd -c src/foo.d -o obj/foo.o -cache` would generate an object and compilation cache for module `foo` under `obj`.   First compilation of `foo.d` generates the cache file, then subsequent compilations of the module reuse it.

Reply:
I'd opt for `-cache` to use a well-know per-user location, and `-cache=path/to/whatever` as optional override.
Since we had that discussion in dub recently (https://github.com/dlang/dub/issues/341), you want to use `$XDG_CACHE_HOME/dmd` as cache directory (or `$HOME/.cache/dmd` as fallback).
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
Windows and OSX obviously having their own folder logic.
`C:\Users\<user>\AppData\Local\dmd\cache` and `~/Library/Caches/dmd` maybe.

Reply:
Where do I put cache by default if all of the above fails? Or maybe I should fail?
XDG_CACHE_HOME is not defined on Ubuntu 16.04 for example.